rllib/algorithms/algorithm.py

```diff
@@ -466,22 +466,31 @@ def __init__(
             # `self.get_default_config()` also returned a dict ->
             # Last resort: Create core AlgorithmConfig from merged dicts.
             if isinstance(default_config, dict):
-                config = AlgorithmConfig.from_dict(
-                    config_dict=self.merge_algorithm_configs(
-                        default_config, config, True
+                if "class" in config:
+                    AlgorithmConfig.from_state(config)
+                else:
+                    config = AlgorithmConfig.from_dict(
+                        config_dict=self.merge_algorithm_configs(
+                            default_config, config, True
+                        )
                     )
-                )
+
             # Default config is an AlgorithmConfig -> update its properties
             # from the given config dict.
             else:
-                config = default_config.update_from_dict(config)
+                if isinstance(config, dict) and "class" in config:
```

Comment:
Actually, let's do the following with this PR:
* Make `AlgorithmConfig` a `Checkpointable` and override the `get/set_state` and `get_ctor_args_and_kwargs` methods first:
```
def get_state(self, *, components=...):
    return [basically the AlgoConfig as a dict]

def set_state(self, state):
    # Here, we can conveniently check, whether keys in `state` exist as properties and override the correct properties as wanted.

def get_ctor_args_and_kwargs(self):
    return ()  # <- empty tuple AlgorithmConfigs are always constructed w/o any args/kwargs
```

Reply:
Then we can do this here:
```
if isinstance(config, dict):
    config = default_config
    config.set_state(config)
# else:
#     keep config as is (it's already a proper AlgorithmConfig object)
```

So, basically, we from here on treat user-provided config dicts as "AlgorithmConfig state dicts".